### PR TITLE
Add basicapp pkg

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/giantswarm/issues/6078

The only difference between `managedservice` is that `managedapp` is pulling chart tarball from our app catalog. 

Please see the spec for more information. 
https://github.com/giantswarm/giantswarm/blob/master/specs/app-catalog-phase-2.md#basic-app-e2e-test

`Service` object checking would follow as next PR.

### CODE REVIEW

File: managedapps/spec.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+type Clients interface {
+	// K8sClient returns a properly configured control plane client for the
+	// Kubernetes API.
+	K8sClient() kubernetes.Interface
+}
+
+// Chart is the chart to test.
+type Chart struct {
+	Name        string
+	URL         string
```
Comment (tomahawk28) [inline managedapps/spec.go]: e2etests would get full URL only. each e2etest on operators should supply the URL.

File: managedapps/legacyresource/resource.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("ensured deletion of release %#q", name))
+
+	return nil
+}
+
+func (r *Resource) Install(name, url, values string, conditions ...func() error) error {
+	ctx := context.TODO()
+
+	tarballPath, err := r.helmClient.PullChartTarball(ctx, url)
```
Comment (tomahawk28) [inline managedapps/legacyresource/resource.go]: We are pulling chart tarball file from here, and the rest of codes are almost same.

File: managedapps/spec.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+// DaemonSet is a daemonset to be tested.
+type DaemonSet struct {
+	Name        string
+	Namespace   string
+	Labels      map[string]string
+	MatchLabels map[string]string
+	Replicas    int
+}
+
+// Deployment is a deployment to be tested.
```
Comment (tomahawk28) [inline managedapps/spec.go]: We also need to test `Service` but this will be next PR.

**Review by rossf7 — COMMENTED** [review]
Direction is nice but I'd like this to be called `basicapp` like it is in the spec.

The reason is we will use it for https://github.com/giantswarm/giantswarm/issues/6224 as a smoke test for all components we install from app catalogs. Including operators installed by VOO.

For example cluster-operator has no e2e tests at the moment. So if the operator crashes we get paged which causes stress and wastes time.

Name doesn't have to be basicapp if someone has a good idea for a different name.

**Comment by tomahawk28** [comment]
@rossf7 ACK, basicapp naming makes sense.

File: managedapps/legacyresource/error.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package legacyresource
```
Comment (rossf7) [inline managedapps/legacyresource/error.go]: This legacyresource was extracted from e2e-harness. Let's keep it for now. We can refactor later.

File: managedapps/spec.go (hunk 3 above)
Comment (rossf7) [reply managedapps/spec.go]: Yes nice <3 we should totally extend to cover services. But yes makes sense to do in a later PR.

File: managedapps/spec.go (hunk 1 above)
Comment (rossf7) [reply managedapps/spec.go]: So the e2e test passes in the URL but what generates that? 

I think we could have that in e2esetup. We get the SHA from the Circle env var and I think the catalog could be config for the e2e test.

**Review by rossf7 — COMMENTED** [review]
Thanks for renaming. <3 My main question is how we generate the tarball URL that is passed in.

File: managedapps/spec.go (hunk 1 above)
Comment (tomahawk28) [reply managedapps/spec.go]: @rossf7  To list the file in github, I think we need to have github API token. does `e2eseteup` already take into account of that feature?

Comment (rossf7) [reply managedapps/spec.go]: Yes, we already do in this in operator e2e tests to get the release index. So we know the WIP and active version bundles.

e.g. https://github.com/giantswarm/aws-operator/blob/82a58529f7c24b071ef37de26c0681262f86a51d/integration/test/scaling/project.yaml#L11

We can use the same approach.

Comment (tomahawk28) [reply managedapps/spec.go]: Sure, then I agree to have below features inside `e2esetups`
- Listing app catalog *.tar and find a file which contains same hash.

File: basicapp/managedapps.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
+package basicapp
```
Comment (rossf7) [inline basicapp/managedapps.go]: Filename needs to be updated.

File: basicapp/managedapps.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+type Config struct {
+	Clients    Clients
+	HelmClient *helmclient.Client
+	Logger     micrologger.Logger
+
+	App            Chart
+	ChartResources ChartResources
+}
+
+type ManagedApps struct {
```
Comment (rossf7) [inline basicapp/managedapps.go]:
```suggestion
type BasicApp struct {
```

File: basicapp/managedapps.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+		resource:   resource,
+
+		chart:          config.App,
+		chartResources: config.ChartResources,
+	}
+
+	return ms, nil
+}
+
+func (ms *ManagedApps) Test(ctx context.Context) error {
```
Comment (rossf7) [inline basicapp/managedapps.go]:
```suggestion
func (b *BasicApp) Test(ctx context.Context) error {
```

This is my bad but I think just `b` is enough.

File: basicapp/managedapps.go [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+// checkDeployment ensures that key properties of the deployment are correct.
+func (ms *ManagedApps) checkDeployment(expectedDeployment Deployment) error {
+	ds, err := ms.clients.K8sClient().Apps().Deployments(expectedDeployment.Namespace).Get(expectedDeployment.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return microerror.Maskf(notFoundError, "deployment: %#q", expectedDeployment.Name)
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if int32(expectedDeployment.Replicas) != *ds.Spec.Replicas {
```
Comment (rossf7) [inline basicapp/managedapps.go]: This is dumb and again my bad :(

Can we remove replicas from the spec?  We can just check that the ready replicas matches the number in the spec. WDYT?

File: basicapp/spec.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	return nil
+}
+
+// DaemonSet is a daemonset to be tested.
+type DaemonSet struct {
+	Name        string
+	Namespace   string
+	Labels      map[string]string
+	MatchLabels map[string]string
+	Replicas    int
```
Comment (rossf7) [inline basicapp/spec.go]: Same as above. I think we can drop the replicas and just use the value we specify in the chart.

**Review by rossf7 — COMMENTED** [review]
Some minor stuff and then will be good. It would also be nice to tidy up the replica count checks.

File: basicapp/managedapps.go (hunk 8 above)
Comment (tomahawk28) [reply basicapp/managedapps.go]: I think we can compare `.Spec.Replicas` == `.Status.ReadyReplicas`. What do you think?

Comment (tomahawk28) [reply basicapp/managedapps.go]: But I don't like it in general since it would take a while until all pods are counted in `ReadyReplica`.

Comment (rossf7) [reply basicapp/managedapps.go]: Up to you but I'm fine with waiting. For complex apps like Ingress Controller I think its a better test as the pods should be ready.

Comment (tomahawk28) [reply basicapp/managedapps.go]: @rossf7 Added retry mechanism. PTAL.

File: basicapp/basicapp.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+			Namespace:  config.App.Namespace,
+		}
+
+		resource, err = legacyresource.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	ms := &BasicApp{
```
Comment (rossf7) [inline basicapp/basicapp.go]:
```suggestion
	b := &BasicApp{
```

File: basicapp/basicapp.go [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+	{
+		b.logger.LogCtx(ctx, "level", "debug", "message", "waiting for deployed status")
+
+		err = b.resource.WaitForStatus(b.chart.Name, "DEPLOYED")
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		b.logger.LogCtx(ctx, "level", "debug", "message", "chart is deployed")
+	}
```
Comment (rossf7) [inline basicapp/basicapp.go:110]: Newline pls.

File: basicapp/basicapp.go [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+	o := func() error {
+		err := b.checkDeploymentReady(ctx, expectedDeployment)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		return nil
+	}
+
+	off := backoff.NewConstant(30*time.Second, 5*time.Second)
```
Comment (rossf7) [inline basicapp/basicapp.go:193]:
```suggestion
	bo := backoff.NewConstant(30*time.Second, 5*time.Second)
```

Convention is to use b but it collides. WDYT to this?

**Review by rossf7 — APPROVED** [review]
LGTM after some final stuff. Also the PR title needs updating :) Thanks for getting this sorted! 🚀

File: basicapp/basicapp.go (hunk 12 above)
Comment (tomahawk28) [reply basicapp/basicapp.go:193]: bo! nice word 😆

**Pull request merged**